service/src/main/java/com/epam/esm/service/service/certificate/CertificateServiceImpl.java

```diff
@@ -0,0 +1,247 @@
+                    ErrorCodes.CERTIFICATE_NOT_FOUND);
+        } else {
+            return certificates;
+        }
+    }
+
+    private List<Certificate> readBy(EntityFinder<Certificate> entityFinder) throws NotFoundException {
+        List<Certificate> certificates = dao.readBy(entityFinder);
+        if (CollectionUtils.isEmpty(certificates)) {
+            throw new NotFoundException("Requested certificate not found!",
+                    ErrorCodes.CERTIFICATE_NOT_FOUND);
+        } else {
+            return certificates;
+        }
+    }
+
+    private void parsePaginationParameter(CertificateFinder finder, String parameterString, List<String> list) {
+        PaginationParameters parameter = PaginationParameters.getEntryByParameter(parameterString);
+        switch (parameter) {
+            case LIMIT:
+                finder.limit(Integer.parseInt(list.get(0)));
+                break;
+            case OFFSET:
+                finder.offset(Integer.parseInt(list.get(0)));
+                break;
+        }
+    }
+
+    private void parseFindParameter(CertificateFinder finder, String parameterString, List<String> list) {
+        CertificateSearchParameters parameter =
+                CertificateSearchParameters.getEntryByParameter(parameterString);
+        switch (parameter) {
+            case NAME:
+                addToFinder(finder::findByName, list);
+                break;
+            case DESCRIPTION:
+                addToFinder(finder::findByDescription, list);
+                break;
+            case TAG_ID:
+                if (list.size() > 1) {
+                    finder.findByTags(list.stream().mapToInt(Integer::parseInt).toArray());
+                } else {
+                    finder.findByTagId(Integer.parseInt(list.get(0)));
+                }
+                break;
+            case TAG_NAME:
+                if (list.size() > 1) {
+                    finder.findByTags(list.toArray(String[]::new));
+                } else {
+                    finder.findByTag(list.get(0));
+                }
+                break;
+        }
+    }
+
+    private void addToFinder(Consumer<String> consumer, String value) {
+        if (StringUtils.isNotEmpty(value)) {
+            consumer.accept(decodeParam(value));
+        }
+    }
+
+    private void addToFinder(Consumer<String> consumer, List<String> list) {
+        if (CollectionUtils.isNotEmpty(list)) {
+            addToFinder(consumer, list.get(0));
+        }
+    }
+
+    @Lookup
+    @Override
+    public CertificateFinder getFinder() {
+        return null;
+    }
+
+    @Override
+    public List<Certificate> read(MultiValueMap<String, String> params)
+            throws BadRequestException, NotFoundException {
+        CertificateFinder finder = getFinder();
+        for (String key : params.keySet()) {
+            try {
+                if (CollectionUtils.isEmpty(params.get(key))) {
```

Comment:
refactor if-else , separate on methods

validateParameters
sortParameters

need to discuss.

Reply:
refactored